cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst

```diff
@@ -317,6 +317,61 @@ For example, give these arguments when starting the MapReduce through a RESTful
     "dataset.totals.output.partition.key.league" : "nfl"
   }
 
+Incrementally Processing PartitionedFileSets
+============================================
+
+One way to process a partitioned file set is with a repeatedly-running MapReduce program that,
+in each run, reads all partitions that have been added since its previous run. This requires
+that the MapReduce program persists between runs which partitions have already been consumed.
+An easy way is to use the ``BatchPartitionConsumer``, an experimental feature introduced in CDAP 3.1.0.
+Your MapReduce program is responsible for extending this abstract consumer class with methods to
+persist and then read back its state. In this example, the state is persisted to a row in a
+KeyValue Table; however, other types of Datasets can also be used::
+
+  public static class WordCount extends AbstractMapReduce {
+
+    private final BatchPartitionConsumer batchPartitionConsumer = new BatchPartitionConsumer() {
+      private static final String STATE_KEY = "state.key";
+
+      @Nullable
+      @Override
+      protected byte[] readBytes(DatasetContext datasetContext) {
+        return ((KeyValueTable) datasetContext.getDataset("consumingState")).read(STATE_KEY);
+      }
+
+      @Override
+      protected void writeBytes(DatasetContext datasetContext, byte[] stateBytes) {
+        ((KeyValueTable) datasetContext.getDataset("consumingState")).write(STATE_KEY, stateBytes);
+      }
+    };
+
+Then, in the ``beforeSubmit()`` method of the MapReduce, specify the partitioned file set to be used as input::
+
+    @Override
+    public void beforeSubmit(MapReduceContext context) throws Exception {
+      PartitionedFileSet inputRecords =
+        batchPartitionConsumer.getPartitionedFileSetForConsuming(context, "inputRecords");
+      context.setInput("inputRecords", inputRecords);
+      ...
+    }
+
+This will read back the previously persisted state, determine the new partitions to read based upon this
+state, and compute a new state to store in memory until a call to ``persist()``. The dataset it returns
+is instantiated with the set of new partitions to read as input.
+
+To save the state of partition processing, call the consumer's ``persist()`` method. This ensures that the
+next time the MapReduce job runs, it processes only the newly committed partitions::
+
+  @Override
+  public void onFinish(boolean succeeded, MapReduceContext context) throws Exception {
+    if (succeeded) {
+      batchPartitionConsumer.persist(context);
+    }
+    super.onFinish(succeeded, context);
+  }
+
+A limitation of the ``BatchPartitionConsumer`` is that there can't be concurrent runs of the MapReduce job.
```

Comment:
yes, true, then you would have to retry after failure. OK. for now, let's leave it at this.